docs/proposals/checks_as_wheels.md

```diff
@@ -0,0 +1,288 @@
+each agent release we would pick and include in the package the desired version
+of each check in integrations-core by invoking `pip install -r` on a special
+requirements.txt file that would be part of the agent repository. That file
+would likely contain the most recent version of each check wheel package but
+this wouldn’t be enforced, what it counts is that the requirements file would be
+the unique source of truth stating which checks are shipped with which agent.
+
+#### Dependencies
+
+We would leverage Python integrated tools to keep dependencies sane.
+For example, pip would refuse to downgrade a check if this breaks the
+dependencies for another check. See the Upgrade paragraph for what should happen
+at every agent upgrade.
+
+We could also provide a dummy Agent wheel package integrations would depend on:
+this would enforce compatibility with the core agent at install time (enforced
+by pip) instead of at runtime (enforced by the gent itself after reading check
+metadata). This would also make manifest files unnecessary since we would use
+setuptools to handle package metadata.
+
+#### Distribution
+
+Datadog would distribute wheel packages for integrations-core and integrations-extra
+through their own Python Package Repositories. Bare minimal version would be a
+file server available through HTTPS (S3+cloudfront for example): if we use the
+CI to upload the wheel packages and take care of filesystem structure,
+that would be enough to implement a distribution channel.
+
+If we want to leverage distutils capabilities, included commands like `setup upload`
+and provide a full fledged PyPi-like website where people can browse package
+metadata, we need to install specific software (open source or cloud hosted
+projects implementing it).
+
+Being the two repos separated, we could grant checks maintainers push rights on
+the integrations-extra repo and let the CI build and upload the package to the
+PyPi channel dedicated to extra.
+
+Any tool supporting the development process for the checks will go in the
+official PyPi repository instead. For example, the Python package replicating
+the functionalities that actually live in the agent could be pip installable
+from pypi.org directly, without forcing contributors to install the full
+agent package on their laptop.
+
+#### Upgrade path
+_Note: this will affect the “end user” experience._
+
+##### Recommended solution: nuke core dependencies, keep custom checks
+
+We can wrap pip execution in the agent command line so that when users install a
+custom check, pip is actually invoked whit this parameter:
+`--target=/opt/datadog-agent/sdk package_name`
+
+Pros:
+ * Custom checks are still there after an upgrade, whether they were added with
+ the old strategy (python module copied over /etc/dd-agent/checks.d/) or through
+ pip.
+
+Cons:
+ * Users wouldn’t be able to pin specific check versions (this could be a non
```

Comment:
I don't think that it's a requirement. I don't see customers wanting a new agent with an old check. Customers usually want newer integration with an old agent to avoid a big update. When they do upgrade the agent their goal to have all the latest features. Plus we almost never remove a feature from an integration.

I think we can assume customers are only moving forward in their upgrade (and we should encourage them to do so).